Readme.md
# Sojolicious

Sojolicious is a toolkit for the federated social web, containing
plugins for the powerful web framework L<Mojolicious>
written in Perl.


# Synopsis

```perl
  use Mojolicious::Lite;

  # Load Plugins
  plugin 'Webfinger';
  plugin 'PubSubHubbub';
  plugin 'Salmon';

  # Esablish Salmon Endpoints:
  group {
    under '/salmon';
    (any '/:acct/mentioned')->salmon('mentioned');
    (any '/:acct/all-replies')->salmon('all-replies');
    (any '/signer')->salmon('signer');
  };

  # Add pubsubhubbub callback url
  (any '/pubsub')->pubsub;

  hook on_salmon_follow => sub {
    # ... You received a follow request
  };

  hook on_pubsub_content => sub {
    my ($c, $type, $dom) = @_;
    # ... You receive feed information you subscribed to
  };

  app->start;
```

# Goal

Sojolicious mainly focus on support for the
[OStatus](https://github.com/OStatus)
meta protocol and aims for a straight forward implementation of all
surrounding specifications.

Due to the success of Mastodon it is unlikely this path will be
followed in the future.

The design goal was to  make all plugins useful on their own,
as separated building blocks of OStatus, while playing well
with each other. All plugins are application (despite the fact
that they are Mojolicious plugins) and storage agnostic,
providing event driven hooks for usage.

See [nils-diewald.de](https://www.nils-diewald.de/development/sojolicious)
for recent updates.

# Plugins for ...

- [x] [ActivityStreams](http://activitystrea.ms/specs/atom/1.0/) via [XML::Loy::ActivityStreams](https://metacpan.org/pod/XML::Loy::ActivityStreams)
- [x] [Atom](https://www.ietf.org/rfc/rfc4287.txt) via [XML::Loy::Atom](https://metacpan.org/pod/XML::Loy::Atom)
- [x] [Atom-Threading-Extension](https://www.ietf.org/rfc/rfc4685.txt) via [XML::Loy::Atom::Threading](https://metacpan.org/pod/XML::Loy::Atom::Threading)
- [x] [HostMeta](http://tools.ietf.org/html/draft-hammer-hostmeta) via [Mojolicious-Plugin-HostMeta](https://metacpan.org/release/Mojolicious-Plugin-HostMeta)
- [x] [MagicSignatures](http://salmon-protocol.googlecode.com/svn/trunk/draft-panzer-magicsig-01.html) via [Crypt::MagicSignatures::Envelope](https://metacpan.org/release/Crypt-MagicSignatures-Envelope/) and [Crypt::MagicSignatures::Key](https://metacpan.org/release/Crypt-MagicSignatures-Key/)
- [ ] [OStatus](https://github.com/OStatus)
- [ ] [PortableContacts](http://portablecontacts.net/draft-spec.html)
- [x] [PubSubHubbub](https://github.com/pubsubhubbub) via [Mojolicious::Plugin::PubSubHubbub](https://metacpan.org/release/Mojolicious-Plugin-PubSubHubbub/)
- [ ] [Salmon](https://github.com/salmon-protocol)
- [x] [XRD](http://docs.oasis-open.org/xri/xrd/v1.0/xrd-1.0.html) via [Mojolicious::Plugin::XRD](https://metacpan.org/release/Mojolicious-Plugin-XRD)
- [x] [Webfinger](http://code.google.com/p/webfinger/wiki/WebFingerProtocol) via [Mojolicious::Plugin::WebFinger](https://metacpan.org/release/Mojolicious-Plugin-WebFinger)

There were plans to expand the scope to other social protocols later,
for example [OExchange](http://www.oexchange.org/spec/).

# Acknowledgement

- **ActivityStreams** was developed by Martin Atkins, Will Norris, Chris Messina, Monica Wilkinson, and Rob Dolin.
- **Atom** was developed by Mark Nottingham and Robert Sayre.
- **Atom** Threading Extensions> was developed by James M. Snell.
- **HostMeta** was developed by  Eran Hammer-Lahav and Blaine Cook.
- **MagicSignatures** was developed by John Panzer, Ben Laurie, and Dirk Balfanz.
- **Mojolicious** is written by Sebastian Riedel.
- **OStatus** was developed by Evan Prodromou, Brion Vibber, James Walker, and Zach Copley.
- **PortableContacts** was developed by Joseph Smarr.
- **PubSubHubbub** was developed by Brad Fitzpatrick, Brett Slatkin, and Martin Atkins.
- **Salmon** was developed by John Panzer.
- **Webfinger** was developed by Brad Fitzpatrick, Eran Hammer-Lahav, Blaine Cook, John Panzer, and Joe Gregorio.
- **XRD** was developed by Eran Hammer-Lahav and Will Norris.

... just to name the persons officially responsible for maintaining the code and the specifications.
Thanks to all contributors of these projects as well!

And thanks to all implementors of these specifications for inspiring code (which is referenced in the sourcecode).

Participation on conferences was supported by the BMBF-project [Linguistic Networks](http://project.linguistic-networks.net/).

# Where to learn more?

[http://mojolicio.us](http://mojolicio.us),[http://ostatus.org/](https://github.com/OStatus).

# ... and?

Copyright (C) 2011-2013, L<Nils Diewald|http://nils-diewald.de/>.

This program is free software, you can redistribute it
and/or modify it under the same terms as Perl.


Changes
0.30 2016-08-07
        - Removed WebFinger, as it is a separated
	  module now.
	- Removed TagHelpers::Pagination, as it is a separated
	  module now.
	- Removed Util::Callback, as it is a separated
	  module now.
	- Updated dependencies.

0.29 2013-03-31
        - Separated M::P::HostMeta.

0.28 2013-03-24
        - Separated M::P::XRD.

0.27 2013-03-08
        - All M::P::XML modules are now part of XML::Loy.
          They will be removed in the next release.

0.26 2013-02-28
        - Made M::P::MagicSignatures::Key a separated distribution
          as Crypt::MagicSignatures::Key.
        - Made M::P::MagicSignatures::Envelope a separated distribution
          as Crypt::MagicSignatures::Envelope.
	- Change the "no dependencies other than perl and mojo" rule
          in the readme.
	- Add -secure flag to Hammer stack discovery for fallback prevention.
	- Change readme to use Mojolicious::Lite example.

0.25 2012-11-14
        - Added M::P::T::Pagination.

0.24 2012-11-10
        - Made M::P::Util::Endpoint a separated distribution.
        - Made M::P::PubSubHubbub a separated distribution.
        - Made M::P::CHI a separated distribution.
        - Made M::P::Oro a separated distribution.

0.23 2012-09-26
	- Update Sojolicious for Mojolicious 3.43.

0.22 2012-07-04
        - Added 'on_oro_connect' event to M::P::Oro.
	- Update Sojolicious for Mojolicious 3.0.

0.21 2012-05-26
        - Moved Sojolicious::Oro to DBIx::Oro.
          Version numbers will now be independent.

0.20 2012-05-20
        - Small optimizations to Oro.
        - Deleted deprecated methods from Oro.

0.19 2012-05-01
        - Added "rel" parameter support to M::P::XRD.
	- Added "resource" parameter support to M::P::HostMeta.
	- Prepared ComplexValues for Hypnotoad compatible init.
	- Small bugfix in LRDD.
	- Added simple CHI plugin.
	- Added native caching option to Oro.
        - Added Config-Parameterization.
        - Added oro_init command.
        - Added endpoints command.

0.18 2012-04-25
        - Added on_connect event to Oro.
        - Made Oro-Plugin work with Hypnotoad.
        - Deleted Waypoints from PortableContacts.
        - Fixed a pretty printer bug in M::P::XML::Base.

0.17 2012-04-03
        - Made Sojolicious::ComplexValues independent from Mojo.
        - Made M::P::Date::RFC3339 independent from Mojo.
        - Lots of bugfixes everywhere.

0.16 2012-04-02
        - Split Oro and Oro::SQLite.
	- Added Oro::SQLite.
	- Added snippet, offsets, and matchinfo support to Oro::SQLite.
	- Added attach and detach support to Oro::SQLite.
	- Added password support to Oro.
	- Added treatment support to Oro.
	- Changed default errors in Oro to be printed.
        - Added explain method to Oro.
        - Lots of bugfixes everywhere.

0.15 2012-02-02
        - Added Mojolicious::Plugin::Oro.
        - Made Sojolicious::Oro thread- and fork-safe.
        - Added restrictions to Sojolicious::Oro.
        - Added Simple Joins to Sojolicious::Oro.
	- Removed Pager support from Sojolicious::Oro.
	- Added distinct restriction to Sojolicious::Oro.
	- Added multiple insertions with default to
	  Sojolicious::Oro.
	- Multiple bugfixes to Sojolicious::Oro,
          Sojolicious::ComplexValues, and M::P::Oro.
        - Added operator support for Sojolicious::Oro.

0.14 2011-12-30
        - Added Sojolicious::ComplexValues.
        - Added Transactions to Sojolicious::Oro.
        - Added table method to Sojolicious::Oro and
          improved transactions and multiple insertions.
        - Added ArbitraryBase.

0.13 2011-12-15
        - Made the multiple inserts in Sojolicious::Oro
          independent from Transactions.
        - Half working version of MagicSignatures.
        - Added LRDD support.
        - Fundamental changes to Util::Endpoint.
        - Small improvements to Hostmeta.
        - Bugfixes to Util::Endpoint.

0.12 2011-11-09
        - Added Oro.

0.11 2011-10-30
        - Added GeoRSS (Simple) support.
        - Added CORS support.
        - Fixed lots of bugs in PubSubHubbub.
        - Added PubSubHubbub test suite.

0.10 2011-10-01
        - Refurbished SerialXML and renamed to XML.
        - Added Mojolicious::Plugin::XML namespace.
        - Added lots of tests.
        - Changed everything to support new XML module.

0.09 2011-09-27
        - Refurbished XML::Serial and renamed to SerialXML.
        - Added Atom-Threading.

0.08 2011-09-24
        - Added separate Util::Base64url plugin.
        - First working version of MagicSignatures.

0.07 2011-08-29
        - Added separate Util::Endpoint plugin.

0.06 2011-08-18
        - Added preliminary PortableContacts support.

0.05 2011-08-17
        - Added OStatus skeleton.

0.04 2011-08-10
        - Added ActivityStreams support.
        - Added Date::RFC3339.

0.03 2011-08-06
        - Added Atom support.

0.02 2011-06-09
        - Added preliminary PubSubHubbub support.

0.01 2011-06-02
        - Initial submission to GitHub.


Makefile.PL
#!/usr/bin/env perl
use 5.010001;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    VERSION_FROM => 'Readme.pod',
    NAME         => 'Sojolicious',
    AUTHOR       => 'Nils Diewald',
    PREREQ_PM => {
        'Mojolicious' => 3.91,
        'DBIx::Oro'   => 0.31,
	'XML::Loy'    => 0.44,
        'Mojolicious::Plugin::Util::Endpoint' => 0.19,
        'Mojolicious::Plugin::Util::Callback' => 0.04,
        'Mojolicious::Plugin::PubSubHubbub'   => 0.16,
	'Mojolicious::Plugin::XRD'            => 0.17,
	'Mojolicious::Plugin::HostMeta'       => 0.21,
        'Mojolicious::Plugin::Oro'            => 0.07,
        'Mojolicious::Plugin::CHI'            => 0.13,
        'Mojolicious::Plugin::XML::Loy'       => 0.14,
        'Mojolicious::Plugin::TagHelpers::Pagination' => 0.06,
        'Crypt::MagicSignatures::Key'         => 0.14,
	'Crypt::MagicSignatures::Envelope'    => 0.10,
	'Test::More'  => 0
    },
    test => {TESTS => 't/*.t t/*/*.t t/*/*/*.t t/*/*/*/*.t'}
);

